src/jdk.jdeps/share/classes/com/sun/tools/jnativescan/RestrictedMethodFinder.java

```diff
@@ -0,0 +1,150 @@
+        classesToScan.forEach((_, info) -> {
+            ClassModel classModel = info.model();
+            List<RestrictedUse> perClass = new ArrayList<>();
+            classModel.methods().forEach(methodModel -> {
+                if (methodModel.flags().has(AccessFlag.NATIVE)) {
+                    perClass.add(new NativeMethodDecl(MethodRef.ofModel(methodModel)));
+                } else {
+                    Set<MethodRef> perMethod = new HashSet<>();
+                    methodModel.code()
+                        .ifPresent(code -> {
+                            code.forEach(e -> {
+                                switch (e) {
+                                    case InvokeInstruction invoke -> {
+                                        if (isRestrictedMethod(invoke.method())) {
+                                            perMethod.add(MethodRef.ofMethodRefEntry(invoke.method()));
+                                        }
+                                    }
+                                    default -> {
+                                    }
+                                }
+                            });
+                        });
+                    if (!perMethod.isEmpty()) {
+                        perClass.add(new RestrictedMethodRefs(MethodRef.ofModel(methodModel),
+                                Set.copyOf(perMethod)));
+                    }
+                }
+            });
+            if (!perClass.isEmpty()) {
+                ScannedModule scannedModule = new ScannedModule(info.jarPath(), info.moduleName());
+                restrictedMethods.computeIfAbsent(scannedModule, _ -> new HashMap<>())
+                        .put(classModel.thisClass().asSymbol(), perClass);
+            }
+        });
+        return restrictedMethods;
+    }
+
+    private boolean isRestrictedMethod(MemberRefEntry method) throws JNativeScanFatalError {
+        return switch (method) {
+            case MethodRefEntry mre ->
+                    isRestrictedMethod(mre.owner().asSymbol(), mre.name().stringValue(), mre.typeSymbol());
+            case InterfaceMethodRefEntry mre ->
+                    isRestrictedMethod(mre.owner().asSymbol(), mre.name().stringValue(), mre.typeSymbol());
+            default -> throw new IllegalStateException("Unexpected type: " + method);
+        };
+    }
+
+    private boolean isRestrictedMethod(ClassDesc owner, String name, MethodTypeDesc type) throws JNativeScanFatalError {
+        return CACHE.computeIfAbsent(new MethodRef(owner, name, type), methodRef -> {
+            if (methodRef.owner().isArray()) {
+                // no restricted methods in arrays atm, and we can't look them up since they have no class file
+                return false;
+            }
+            Optional<ClassResolver.Info> info = systemClassResolver.lookup(methodRef.owner());
+            if (!info.isPresent()) {
+                return false;
```

Comment:
Is this just `false` or maybe a warning that a certain owner could not be resolved (maybe if running with enough debug options) ?

Reply:
Yes, thought about that yesterday as well. Good catch

Reply:
Thinking a bit more: we also use the optional being empty to know if the owner is not a system class, so I think this code here is what we want. We can however throw an exception if a class is not found in `SystemClassResolver::lookup`. Since, if a class is inside a system module package, it should be found as well (unless the user sets the release to the wrong version).